<!doctype html>
<html>
  <head>
  <title>
    
      SNP Calling | Blogs of Yingwei Guo
    
  </title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta charset="utf-8">
  <link rel="stylesheet" href="/assets/css/main.css">
  <link rel="stylesheet" href="/assets/css/syntax.css">
  <!-- Use Atom -->
  <link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="Blogs of Yingwei Guo" />
  <!-- Use RSS-2.0 -->
  <!--<link href="/rss-feed.xml" type="application/rss+xml" rel="alternate" title="Blogs of Yingwei Guo | Blogs"/>
  //-->
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Code+Pro">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Quattrocento+Sans">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
  <script type="text/javascript" async
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML">
    MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [['$', '$'], ['\\(', '\\)']]
        }
      });
  </script>
  <!-- Google Analytics -->
  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', '', 'auto');
  ga('send', 'pageview');
</script>

  <!-- Use Jekyll SEO plugin -->
  <!-- Begin Jekyll SEO tag v2.7.1 -->
<title>SNP Calling | Blogs of Yingwei Guo</title>
<meta name="generator" content="Jekyll v4.2.1" />
<meta property="og:title" content="SNP Calling" />
<meta name="author" content="Yingwei Guo" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="01.QC 02.REFERENCE INDEX 03.MAPPING 04.CALLING 05.FILTER 06.CONCATENATE 07.ANNOTATION 08.PRUNE 09.PHASE 10.IMPUTATION" />
<meta property="og:description" content="01.QC 02.REFERENCE INDEX 03.MAPPING 04.CALLING 05.FILTER 06.CONCATENATE 07.ANNOTATION 08.PRUNE 09.PHASE 10.IMPUTATION" />
<link rel="canonical" href="http://localhost:4000/snp-calling" />
<meta property="og:url" content="http://localhost:4000/snp-calling" />
<meta property="og:site_name" content="Blogs of Yingwei Guo" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2021-12-12T00:00:00+08:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="SNP Calling" />
<script type="application/ld+json">
{"author":{"@type":"Person","name":"Yingwei Guo"},"@type":"BlogPosting","headline":"SNP Calling","dateModified":"2021-12-12T00:00:00+08:00","datePublished":"2021-12-12T00:00:00+08:00","description":"01.QC 02.REFERENCE INDEX 03.MAPPING 04.CALLING 05.FILTER 06.CONCATENATE 07.ANNOTATION 08.PRUNE 09.PHASE 10.IMPUTATION","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/snp-calling"},"url":"http://localhost:4000/snp-calling","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->

</head>

  <body>
    <div class="container">
      <header class="header">
  <h3 class="header-title">
    <a href="/">Blogs of Yingwei Guo</a>
    <small class="header-subtitle">Blogs</small>
    <div class="menu">
  <nav class="menu-content">
    
      <a href="/about.html">About</a>
    
      <a href="/writing.html">Writing</a>
    
      <a href="/contact.html">Contact</a>
    
  </nav>
  <nav class="social-icons">
    
  
  
    <a href="https://www.github.com/guoyingwei6" target="_blank"><i class="fa fa-github" aria-hidden="true"></i></a>
  

  
  
    <a href="https://twitter.com/" target="_blank"><i class="fa fa-twitter" aria-hidden="true"></i></a>
  

  
  
    <a href="http://www.google.com" target="_blank"><i class="fa fa-linkedin" aria-hidden="true"></i></a>
  

  
  
    <a href="mailto:guoyingwei6@gmail.com" target="_blank"><i class="fa fa-envelope" aria-hidden="true"></i></a>
  

  
  
    <a href="/feed.xml"><i class="fa fa-rss-square" aria-hidden="true"></i></a>
  

  </nav>
</div>

  </h3>
</header>

      <div class="content-container">
        <h1>
  SNP Calling
</h1>

<article>
  <p>01.QC
02.REFERENCE INDEX
03.MAPPING
04.CALLING
05.FILTER
06.CONCATENATE
07.ANNOTATION 
08.PRUNE
09.PHASE
10.IMPUTATION</p>

<h2 id="01qc">01.QC</h2>

<p><code class="language-plaintext highlighter-rouge">/stor9000/apps/users/NWSUAF/2016050001/software/fastp -i ${1}_1.fq.gz -I ${1}_2.fq.gz -o ${1}_1.clean.fq.gz -O ${1}_2.clean.fq.gz -h ${1}.html</code></p>

<p><code class="language-plaintext highlighter-rouge">-n 5 -q 15 -u 40</code>
去除接头序列(adapter);
当测序read中含有的N的含量超过该条read长度比例的5%时，需要去除此对paired reads；
当测序read中含有的低质量（Q ≤15）碱基数超过该条read长度比例的40% 时，需要去除此对paired reads。
<a href="https://github.com/OpenGene/fastp">fastp</a></p>

<h2 id="02reference-index">02.REFERENCE INDEX</h2>

<p>1.Generate the BWA index using BWA. This will create the “.fasta.amb”, “.fasta.ann”, “.fasta.bwt”, “.fasta.pac” and “.fasta.sa” files.
<code class="language-plaintext highlighter-rouge">bwa index GCF_002742125.1_Oar_rambouillet_v1.0_genomic.rename.AddY.fna</code></p>

<p>2.Generate the FASTA file index using samtools. This will create the “.fasta.fai” file.
<code class="language-plaintext highlighter-rouge">samtools faidx GCF_002742125.1_Oar_rambouillet_v1.0_genomic.rename.AddY.fna</code></p>

<p>3.Generate the sequence dictionary using Picard. This will create the “.dict” file.
<code class="language-plaintext highlighter-rouge">/stor9000/apps/appsoftware/BioSoftware/bin/java -jar bin/picard_2.18.17.jar CreateSequenceDictionary R=GCF_002742125.1_Oar_rambouillet_v1.0_genomic.rename.AddY.fna O=GCF_002742125.1_Oar_rambouillet_v1.0_genomic.rename.AddY.dict</code></p>

<h2 id="03mapping">03.MAPPING</h2>
<p>define variable</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>export sm=${1}
export fq1=${2}
export fq2=${3}
export ref=Sheep/reference/Oar4.0/Oar4.0_add_CPY_for_target_seq/Oar4.0_add_CPY.fa
</code></pre></div></div>

<p>mapping</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/Software/bwa-0.7.17/bwa mem \
-t 4 \
-R '@RG\tID:'${sm}'\tLB:'${sm}'\tPL:ILLUMINA\tSM:'${sm} \
${ref} \
${fq1} \
${fq2} \
| /Software/samtools1.9/bin/samtools view -bS -o ${sm}.bam
</code></pre></div></div>
<p>sort</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/stor9000/apps/appsoftware/BioSoftware/bin/java \
-Xmx10g \
-Djava.io.tmpdir=files/tmp \
-Dpicard.useLegacyParser=false \
-jar bin/picard_2.18.17.jar \
SortSam \
-I ${sm}.bam \
-O ${sm}.sort.bam \
-SORT_ORDER coordinate \
-VALIDATION_STRINGENCY LENIENT
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dedup 去除PCR重复
/stor9000/apps/appsoftware/BioSoftware/bin/java \
-Xmx10g \
-Djava.io.tmpdir=files/tmp \
-Dpicard.useLegacyParser=false \
-jar bin/picard_2.18.17.jar \
MarkDuplicates \
-I ${sm}.sort.bam \
-O ${sm}.dedup.bam \
-ASSUME_SORT_ORDER coordinate \
-METRICS_FILE ${sm}.dedup.txt \
-VALIDATION_STRINGENCY LENIENT
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># index bam
/stor9000/apps/users/NWSUAF/2014010784/software/samtools/samtools-1.12/samtools index ${sm}.dedup.bam
# realign Indel区域的重新比对
/stor9000/apps/appsoftware/BioSoftware/bin/java \
-Xmx10g \
-Djava.io.tmpdir=files/tmp \
-jar bin/GATK3.7/GenomeAnalysisTK.jar \
-T RealignerTargetCreator \
-I ${sm}.dedup.bam \
-R ${ref} \
-o ${sm}.intervals

/stor9000/apps/appsoftware/BioSoftware/bin/java \
-Xmx10g \
-Djava.io.tmpdir=files/tmp \
-jar bin/GATK3.7/GenomeAnalysisTK.jar \
-T IndelRealigner \
-I ${sm}.dedup.bam \
-R ${ref} \
-targetIntervals ${sm}.intervals \
-o ${sm}.realign.bam

</code></pre></div></div>

<h2 id="04calling">04.CALLING</h2>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>export sm=${1}
export chr=${3}
export bam=${2}
export ref=Sheep/reference/Oar4.0/Oar4.0_add_CPY_for_target_seq/Oar4.0_add_CPY.fa
bin/gatk-4.1.2.0/gatk \
--spark-runner LOCAL \
--java-options "-Djava.io.tmpdir=files/tmp -Xmx10G" \
**HaplotypeCaller** \
-R ${ref} \
-I ${bam} \
-ERC GVCF \
-L ${chr} \
-O 01.gvcf/${sm}/${sm}.${chr}.gvcf.gz

export chr=${1}
export ref=Sheep/reference/Oar4.0/Oar4.0_add_CPY_for_target_seq/Oar4.0_add_CPY.fa

bin/gatk-4.1.2.0/gatk \
--java-options "-Djava.io.tmpdir=files/tmp -Xmx10G" \
**CombineGVCFs** \
-R ${ref} \
--variant files/snp_calling/04.calling/01.gvcf/GJ01429/GJ01429.chr${chr}.gvcf.gz \
--variant files/snp_calling/04.calling/01.gvcf/GJ01741/GJ01741.chr${chr}.gvcf.gz \
--output 02.merged_gvcf/chr${chr}.gvcf.gz

export chr=${1}
export ref=Sheep/reference/Oar4.0/Oar4.0_add_CPY_for_target_seq/Oar4.0_add_CPY.fa
bin/gatk-4.1.2.0/gatk \
--java-options "-Djava.io.tmpdir=files/tmp -Xmx10G" \
**GenotypeGVCFs** \
-R ${ref} \
--variant 02.merged_gvcf/chr${chr}.gvcf.gz \
--output 03.vcf/chr${chr}.vcf.gz

</code></pre></div></div>
<h2 id="05filter">05.FILTER</h2>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bcftools view \
-v snps \
-e 'QD &lt; 2.0 | QUAL &lt; 30.0 | SOR &gt; 3.0 | FS &gt; 60.0 | MQ &lt; 40.0 | MQRankSum &lt; -12.5 | ReadPosRankSum &lt; -8.0 ' \
-m2 -M2 \
../04.calling/03.vcf/chr${1}.vcf.gz \
-Oz \
-o chr${1}.filtered.vcf.gz


#bcftools view -i 'F_MISSING &lt; 0.1 &amp; MAF &gt; 0.05' vcf -Oz -o filter.vcf

</code></pre></div></div>

</article>

  <span class="post-date">
  Written on
  
  December
  12th,
  2021
  by
  
    Yingwei Guo
  
</span>



  <div class="post-date">Feel free to share!</div>
<div class="sharing-icons">
  <a href="https://twitter.com/intent/tweet?text=SNP Calling&amp;url=/snp-calling" target="_blank"><i class="fa fa-twitter" aria-hidden="true"></i></a>
  <a href="https://www.facebook.com/sharer/sharer.php?u=/snp-calling&amp;title=SNP Calling" target="_blank"><i class="fa fa-facebook" aria-hidden="true"></i></a>
</div>



  <div class="related">
  <h1 >You may also enjoy:</h1>
  
  <ul class="related-posts">
    
      
        
          <li>
            <h3>
              <a href="/muti-test">
                【转】GWAS当中的多重比较问题：Bonferroni校正，FDR控制和置换测试permutation
                <!--<img src="http://localhost:4000/images/">-->
                <!--<small>January 7, 2022</small>-->
              </a>
            </h3>
          </li>
          
        
      
        
          <li>
            <h3>
              <a href="/plink">
                【GWAS】利用PLINK软件进行数据清洗和关联分析
                <!--<img src="http://localhost:4000/images/">-->
                <!--<small>January 2, 2022</small>-->
              </a>
            </h3>
          </li>
          
        
      
        
          <li>
            <h3>
              <a href="/emmax">
                【GWAS】利用EMMAX软件进行GWAS
                <!--<img src="http://localhost:4000/images/">-->
                <!--<small>January 1, 2022</small>-->
              </a>
            </h3>
          </li>
          
        
      
    
  </ul>
</div>




      </div>
      <footer class="footer">
  
  
  
    <a href="https://www.github.com/guoyingwei6" target="_blank"><i class="fa fa-github" aria-hidden="true"></i></a>
  

  
  
    <a href="https://twitter.com/" target="_blank"><i class="fa fa-twitter" aria-hidden="true"></i></a>
  

  
  
    <a href="http://www.google.com" target="_blank"><i class="fa fa-linkedin" aria-hidden="true"></i></a>
  

  
  
    <a href="mailto:guoyingwei6@gmail.com" target="_blank"><i class="fa fa-envelope" aria-hidden="true"></i></a>
  

  
  
    <a href="/feed.xml"><i class="fa fa-rss-square" aria-hidden="true"></i></a>
  

  <div class="footer-description"><a href="/">Blogs of Yingwei Guo | Blogs by Yingwei Guo</a></div>
</footer>

    </div>
  </body>
</html>
